# Building a SQL database
This notebook builds the necessary db files for the project using SQLite3.

Most data is taken from Chicago Data Portal https://data.cityofchicago.org/ using Socrata library.
The API endpoints for the data are:
- Red Light Violations: https://data.cityofchicago.org/resource/spqx-js37.json
- Congestion by Region 2018-Present: https://data.cityofchicago.org/resource/kf7e-cur8.json
- Congestion by Region 2013-2018: https://data.cityofchicago.org/resource/emtn-qqdi.json
- Traffic Crashes: https://data.cityofchicago.org/resource/85ca-t3if.json

Weather data is taken from https://openweathermap.org/weather-data and is saved as csv in data folder

Tables to build:
- daily_violations (one entry for each camera each day with total violations)
- intersection_locations (one entry for each intersection with lat/long)
- intersection_cams (one entry for each intersection with camera_ids)
- signal_crashes (one entry for each intersection crash with many columns)
- cam_locations (one entry for each cam, with lat/long)
- cam_startend (one entry for each cam with start end dates for min/max dates active)
- hourly_congestion (one entry per hour with bus speed averages for each region)
- hourly_weather (one entry per hour with many weather cols)
- region_data (one entry per region with locations and descriptions to place intersections)


## Connecting and Setup

### Required Imports

In [1]:
import pandas as pd
from sodapy import Socrata
#import matplotlib.pyplot as plt
from datetime import datetime
from modules.myfuncs import *
import warnings
import numpy as np
from geopy.geocoders import Nominatim
from scipy.stats import mode, percentileofscore

warnings.filterwarnings('ignore')

### Create/connect to db and build the TABLEs
We create the connection and cursor objects we will use to communicate with our SQLite database.

In [2]:
# Create a db file or open connection
conn = create_connection('database/rlc.db')  # function I created in myfuncs file
c = conn.cursor()
#conn.close()

sqlite3 version: 2.6.0
connected to database/rlc.db


### Set up the Socrata client
The Chicago Data Portal, which contains most of my data used in this project, uses the Socrata software which can be accessed through Python's sodapy library.
Here we create a client, which we will use to query the data at the portal.

The data API is at [data.cityofchicago.org](data.cityofchicago.org)
The individual db endpoints are found by browsing the site.

In [3]:
# Unauthenticated client only works with public data sets. Note 'None'
# in place of application token, and no username or password:

url = "data.cityofchicago.org"
client = Socrata(url, None)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(data.cityofchicago.org,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

# Build the TABLEs

For every TABLE
- Use a Socrata client query to get all relevant data
- Preprocess data as needed
- Create Table

Our data
- rlc_cam is up to 1M redlight cams from 2015 to 2020
- crash_data is up to 1M crashes from 2015 to 2020
- traffic_data is up to 10M from 2015 to 2020
- wx_data is a csv file from [openweathermap.org](openweathermap.org)

Weather data is taken from csv in data folder

## 1) Build intersection_chars TABLE from int_df

This data is created by me.  It is a dictionary contained in this repository under the file 'int_chars.py'.

To create this file, I went through all 180+ intersections with red light cameras.  I cross referenced it with a map at https://data.cityofchicago.org/Transportation/Average-Daily-Traffic-Counts-Map/pf56-35rv and google maps to compile the following data.

- roads (list): of road segments as identified in average-daily-traffic-counts db in link above.  Used to determine volume of traffic.
- protected_turn (int): How any of the left turns are protected (left turn arrow).
- total_lanes (int): Count of total lanes.  If a road has one lane for all directions N/E/S/W bound traffic, that counts as 4.  Rangees from 3 to 14 lanes.
- medians (int): Count of physical median barriers that extend up to intersection.
- exit (int): 0 if no exit.  1 if exit on/off ramp within 100m of center of intersection.  Traffic flow is affected by proximity to exit.
- split (int): 1 if it is a divided boulevard (common in Chicago) where divided lanes are split by traffic signals in median. Look at examples on google map.
- way (int): directions of traffic flow. A 4 way intersection might be NESW.
- underpass (int): number of ways that have an underpass extending up to the intersection.  These are notoriously bad intersections in Chicago.
- no_left (int): number of no left turn signs.  Usually with smaller streets onto larger roads or high volumne intersections.
- angled (int): 1 if angle between two 2way roads is greater than 30 degrees (used 1/2/sqrt(3) rule to measure.
- triangle (int): 1 if three 2way roads meet intersect or form a triangle where all 3 roads <50m
- one_way (int): number of 1 way directions.
- turn_lanes (int): how many directions have physical and identified turn lanes for left hand turns.
- lat (float): latitude of center of inersection.  
- long (float): longitude 
- rlc (int): 1 for red light camera is present
- intersection (str): name of intersection as defined in signal_crashes table in db
- daily_traffic (int): volume of daily traffic through intersection.  Sum of incoming roads from roads list.

### Preprocess int_df
Import the dictionary and convert to DataFrame which will be written as Table in my db

In [4]:
from modules.int_chars import *
import pandas as pd

int_chars.keys()
int_df = pd.DataFrame.from_dict(int_chars, orient='index')
int_df['intersection'] = int_chars.keys()
int_df.isna().sum()

roads             0
protected_turn    0
total_lanes       0
medians           0
exit              0
split             0
way               0
underpass         0
no_left           0
angled            0
triangle          0
one_way           0
turn_lanes        0
lat               0
long              0
rlc               0
intersection      0
dtype: int64

For now, we will only use intersections with rlc of 1.  I may later add intersections without rlc to identify crash characterisics.

In [5]:
int_df = int_df[int_df['rlc']==1]  # I entertained adding additional non-camera intersections
int_df.columns

Index(['roads', 'protected_turn', 'total_lanes', 'medians', 'exit', 'split',
       'way', 'underpass', 'no_left', 'angled', 'triangle', 'one_way',
       'turn_lanes', 'lat', 'long', 'rlc', 'intersection'],
      dtype='object')

They were read in as non_null objects.  Would like to cast them before creating a table

In [6]:
cols_toint = ['protected_turn', 'total_lanes', 'medians', 'exit', 'split',
       'way', 'underpass', 'no_left', 'angled', 'triangle', 'one_way',
       'turn_lanes', 'rlc']
cols_tofloat = ['lat', 'long',]

int_df[cols_toint] = int_df[cols_toint].astype(int)
int_df[cols_tofloat] = int_df[cols_tofloat].astype(float)
int_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 182 entries, 111TH AND HALSTED to WESTERN AND TOUHY
Data columns (total 17 columns):
roads             182 non-null object
protected_turn    182 non-null int64
total_lanes       182 non-null int64
medians           182 non-null int64
exit              182 non-null int64
split             182 non-null int64
way               182 non-null int64
underpass         182 non-null int64
no_left           182 non-null int64
angled            182 non-null int64
triangle          182 non-null int64
one_way           182 non-null int64
turn_lanes        182 non-null int64
lat               182 non-null float64
long              182 non-null float64
rlc               182 non-null int64
intersection      182 non-null object
dtypes: float64(2), int64(13), object(2)
memory usage: 25.6+ KB


In [7]:
int_df.head()  # verify my table data before commit

roads  protected_turn  \
111TH AND HALSTED                      [28 West, 11600 South]               2   
115TH AND HALSTED                     [714 West, 11600 South]               4   
119TH AND HALSTED                     [446 West, 11600 South]               4   
31ST AND CALIFORNIA                   [2825 West, 3026 South]               2   
31ST ST AND MARTIN LUTHER KING DRIVE   [440 East, 3030 South]               2   

                                      total_lanes  medians  exit  split  way  \
111TH AND HALSTED                               6        2     0      0    4   
115TH AND HALSTED                               6        2     0      0    4   
119TH AND HALSTED                               6        2     0      0    4   
31ST AND CALIFORNIA                             6        0     0      0    4   
31ST ST AND MARTIN LUTHER KING DRIVE           10        2     0      1    4   

                                      underpass  no_left  angled  triangle  \
111TH AND HALSTED                             0        0       1         0   
115TH AND HALSTED                             0        0       0         0   
119TH AND HALSTED                             0        0       0         0   
31ST AND CALIFORNIA                           0        0       0         0   
31ST ST AND MARTIN LUTHER KING DRIVE          0        2       0         0   

                                      one_way  turn_lanes        lat  \
111TH AND HALSTED                           0           2  41.692362   
115TH AND HALSTED                           0           4  41.685089   
119TH AND HALSTED                           0           4  41.677774   
31ST AND CALIFORNIA                         0           4  41.837424   
31ST ST AND MARTIN LUTHER KING DRIVE        0           0  41.838441   

                                           long  rlc  \
111TH AND HALSTED                    -87.642423    1   
115TH AND HALSTED                    -87.642094    1   
119TH AND HALSTED                    -87.641930    1   
31ST AND CALIFORNIA                  -87.695022    1   
31ST ST AND MARTIN LUTHER KING DRIVE -87.617338    1   

                                                              intersection  
111TH AND HALSTED                                        111TH AND HALSTED  
115TH AND HALSTED                                        115TH AND HALSTED  
119TH AND HALSTED                                        119TH AND HALSTED  
31ST AND CALIFORNIA                                    31ST AND CALIFORNIA  
31ST ST AND MARTIN LUTHER KING DRIVE  31ST ST AND MARTIN LUTHER KING DRIVE

### Add daily traffic volume
Now bring in my count information using my roads list associated with each intersection
This data is also from the data portal.  A survey was done in 2013 for two weeks.  Traffic flow was recoreding at many points around the city.  We make the assumption that the traffic patterns at that time are at least somewhat consistent with the more current patterns and can be used as an estimation of how busy the intersections is on a typical day.

In [8]:
daily_traffic = client.get("pfsx-4n4m", 
                     limit=2000,
                    )

daily_traffic = pd.DataFrame.from_records(daily_traffic) # Convert to pandas DataFrame

In [9]:
daily_traffic.info()
cols_tokeep = ['traffic_volume_count_location_address', 'total_passing_vehicle_volume',]
daily_traffic = daily_traffic[cols_tokeep]

daily_traffic.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1279 entries, 0 to 1278
Data columns (total 15 columns):
id                                             1279 non-null object
traffic_volume_count_location_address          1279 non-null object
street                                         1279 non-null object
date_of_count                                  1279 non-null object
total_passing_vehicle_volume                   1279 non-null object
vehicle_volume_by_each_direction_of_traffic    1279 non-null object
latitude                                       1279 non-null object
longitude                                      1279 non-null object
location                                       1279 non-null object
:@computed_region_rpca_8um6                    1266 non-null object
:@computed_region_vrxf_vc4k                    1266 non-null object
:@computed_region_6mkv_f3dw                    1279 non-null object
:@computed_region_bdys_3d7i                    1265 non-null object
:@compute

traffic_volume_count_location_address total_passing_vehicle_volume
0                             5838 West                         7100
1                              320 East                         8600
2                             1730 East                        53500
3                              125 East                          700
4                             2924 East                         4200

In [10]:
daily_traffic.total_passing_vehicle_volume = daily_traffic.total_passing_vehicle_volume.astype(int)

Combine my characteristics with my daily_traffic by looking up traffic volume from daily_traffic df.

In [11]:
int_df.roads.head()

111TH AND HALSTED                        [28 West, 11600 South]
115TH AND HALSTED                       [714 West, 11600 South]
119TH AND HALSTED                       [446 West, 11600 South]
31ST AND CALIFORNIA                     [2825 West, 3026 South]
31ST ST AND MARTIN LUTHER KING DRIVE     [440 East, 3030 South]
Name: roads, dtype: object

In [12]:
def look_up_roads(road_list):
    '''
    Look up function to get the values and return the total
            Parameters:
                roads (list): road segment list for intersection
            Returns:
                total (int): combined traffic volume of every road in roads list.
    '''
    total = 0  
    for road in road_list:
        count = daily_traffic[daily_traffic['traffic_volume_count_location_address']==road]['total_passing_vehicle_volume'].values[0]
        total += count
    return total

int_df['daily_traffic'] = int_df['roads'].apply(look_up_roads)
int_df.drop(columns=['roads'], inplace=True)

In [13]:
int_df.head()

protected_turn  total_lanes  medians  \
111TH AND HALSTED                                  2            6        2   
115TH AND HALSTED                                  4            6        2   
119TH AND HALSTED                                  4            6        2   
31ST AND CALIFORNIA                                2            6        0   
31ST ST AND MARTIN LUTHER KING DRIVE               2           10        2   

                                      exit  split  way  underpass  no_left  \
111TH AND HALSTED                        0      0    4          0        0   
115TH AND HALSTED                        0      0    4          0        0   
119TH AND HALSTED                        0      0    4          0        0   
31ST AND CALIFORNIA                      0      0    4          0        0   
31ST ST AND MARTIN LUTHER KING DRIVE     0      1    4          0        2   

                                      angled  triangle  one_way  turn_lanes  \
111TH AND HALSTED                          1         0        0           2   
115TH AND HALSTED                          0         0        0           4   
119TH AND HALSTED                          0         0        0           4   
31ST AND CALIFORNIA                        0         0        0           4   
31ST ST AND MARTIN LUTHER KING DRIVE       0         0        0           0   

                                            lat       long  rlc  \
111TH AND HALSTED                     41.692362 -87.642423    1   
115TH AND HALSTED                     41.685089 -87.642094    1   
119TH AND HALSTED                     41.677774 -87.641930    1   
31ST AND CALIFORNIA                   41.837424 -87.695022    1   
31ST ST AND MARTIN LUTHER KING DRIVE  41.838441 -87.617338    1   

                                                              intersection  \
111TH AND HALSTED                                        111TH AND HALSTED   
115TH AND HALSTED                                        115TH AND HALSTED   
119TH AND HALSTED                                        119TH AND HALSTED   
31ST AND CALIFORNIA                                    31ST AND CALIFORNIA   
31ST ST AND MARTIN LUTHER KING DRIVE  31ST ST AND MARTIN LUTHER KING DRIVE   

                                      daily_traffic  
111TH AND HALSTED                             43100  
115TH AND HALSTED                             42500  
119TH AND HALSTED                             41800  
31ST AND CALIFORNIA                           41100  
31ST ST AND MARTIN LUTHER KING DRIVE          36500

### Create intersection_chars TABLE

In [14]:
make_table(int_df, 'intersection_chars', c, conn)  # function from import

[('intersection_chars',)]


## 2) Build daily_violations TABLE  from rlc_df

This table will hold the red light camera violations data.  It will include daily violations of every red light camera.
This is a large dataset.  (320 cameras with daily violations over the past 5 years)

### Query red light violations
Use the API endpoint to get data with Socrata query.

In [15]:
# Red light violations
# Takes several minutes to run and holds about 500mb in memory to build

# First 1000000 results, returned as JSON from API / converted to Python list of dictionaries by sodapy
rlc_df = client.get("spqx-js37", #speed cams are at 'hhkd-xvj4' if you want to investigate?
                     #where='violation_date > 01-01-2020',
                     where='violation_date > \'2016-01-01T00:00:00.000\'',
                     limit=10000000,
                    )

rlc_df = pd.DataFrame.from_records(rlc_df) # Convert to pandas DataFrame

### Preprocess Red Light Camera Data

Data Columns of interest (from API docs):

INTERSECTION -
Intersection of the location of the red light enforcement camera(s). There may be more than one camera at each intersection. Plain Text

CAMERA ID -
A unique ID for each physical camera at an intersection, which may contain more than one camera. Plain Text

ADDRESS	-
The address of the physical camera (CAMERA ID). The address may be the same for all cameras or different, based on the physical installation of each camera. Plain Text

VIOLATION DATE -
The date of when the violations occurred. NOTE: The citation may be issued on a different date. Date & Time

VIOLATIONS - 
Number of violations for each camera on a particular day. Number

LATITUDE -
The latitude of the physical location of the camera(s) based on the ADDRESS column. Geocoded using the WGS84. Number

LONGITUDE -
The longitude of the physical location of the camera(s) based on the ADDRESS column. Geocoded using the WGS84.
Number

#### Investigate violation data

In [16]:
rlc_df.info()
rlc_df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 487319 entries, 0 to 487318
Data columns (total 10 columns):
intersection      487319 non-null object
camera_id         487203 non-null object
address           487319 non-null object
violation_date    487319 non-null object
violations        487319 non-null object
x_coordinate      461894 non-null object
y_coordinate      461894 non-null object
latitude          461894 non-null object
longitude         461894 non-null object
location          461894 non-null object
dtypes: object(10)
memory usage: 37.2+ MB


intersection          0
camera_id           116
address               0
violation_date        0
violations            0
x_coordinate      25425
y_coordinate      25425
latitude          25425
longitude         25425
location          25425
dtype: int64

#### Drop nan values and unnecessary columns
We see that we have all text/non-null objects.  Need to convert first before manipulating for preprocess.

There are a fair number of missing locations/lat/long.  Hope to be able to replace those missing values.
This represents a large enough portion of dataset that we should look them up.

The na values for camera_id will have to be dropped, since we don't know what they are.

We will not be using x andy y_coordinate, so we drop those.  We will also drop location.  We already have lat long in other columns.

In [17]:
#client_df.dropna(subset=['camera_id']).isna().sum()
try:
    # put this is a try in case we run it twice, it will skip it.
    rlc_df.dropna(subset=['camera_id'], inplace=True)
    
    # drop xy coord and location columns
    rlc_df = rlc_df.drop(columns=['x_coordinate', 'y_coordinate', 'location'], index=1)
except:
    pass



rlc_df.isna().sum()

intersection          0
camera_id             0
address               0
violation_date        0
violations            0
latitude          25420
longitude         25420
dtype: int64

In [ ]:
rlc_df.intersection.sort_values().unique()

#### Fix intersection errors identified in EDA

In [ ]:
# fix a specific intersection naming problem
# both 'NORTHWEST HIGHWAY AND FOSTER' and 'FOSTER AND NORTHWEST HIGHWAY' exist.  Only one should be used.  They are same.
# Also happens with "MILWAUKEE AND CENTRAL" / "CENTRAL AND MILWAUKEE"
rlc_df['intersection'] = rlc_df['intersection'].apply(lambda x: 'MILWAUKEE AND CENTRAL' if x=='CENTRAL AND MILWAUKEE' else x)
rlc_df['intersection'] = rlc_df['intersection'].apply(lambda x: 'FOSTER AND NORTHWEST HIGHWAY' if x=='NORTHWEST HIGHWAY AND FOSTER' else x)

#### Manipulate datatypes for preprocessing. 

In [ ]:
rlc_df['violations'] = rlc_df['violations'].astype(int)
rlc_df['latitude'] = rlc_df['latitude'].astype(float)
rlc_df['longitude'] = rlc_df['longitude'].astype(float)
rlc_df['violation_date'] = pd.to_datetime(rlc_df['violation_date'])
rlc_df['month'] = rlc_df['violation_date'].apply(lambda x: int(x.month))
rlc_df['day'] = rlc_df['violation_date'].apply(lambda x: int(x.day))  # fixed from dat to day!

rlc_df['weekday'] = rlc_df['violation_date'].apply(lambda x: int(datetime.weekday(x)))
rlc_df['year'] = rlc_df['violation_date'].apply(lambda x: int(x.year))

### Create daily_violations TABLE 

In [ ]:
make_table(rlc_df, 'daily_violations', c, conn)

## 3) Build cam_locations and cam_startend TABLEs
Build tables from cam_locs AND TABLE from cam_startend.
We wll bring in camera locations from data portal and identify the start and end date for each camera.  We will use the start end dates to develop a natural experiment where we identify which cameras were turned on or off during our study timeframe.

### Camera location preprocessing
Make a df with info for each camera
Will contain the following:
- camera_id
- location
- start date (when was the camera turned on)
- end date (when was the camera turned off)

In [ ]:
cam_df = rlc_df.copy()
cam_df['start'] = cam_df['camera_id'].apply(lambda x: None)
cam_df['end'] = cam_df['camera_id'].apply(lambda x: None)

In [ ]:
cam_start = cam_df.groupby(['camera_id'])['violation_date'].min().reset_index()
cam_end = cam_df.groupby(['camera_id'])['violation_date'].max().reset_index()

cam_startend = cam_start.copy()

#print(cam_end[cam_end['camera_id']=='1503'].values[0][1])  # for testing output
cam_startend['end'] = cam_start['camera_id'].apply(lambda x: cam_end[cam_end['camera_id']==x].values[0][1])

cam_startend.rename(columns={"violation_date": "start"}, inplace=True)
                                                   
print('NA values in cam_startend:', cam_startend.isna().sum(), end='\n\n', sep='\n')

print('Describe cam_startend:', cam_startend.describe(), end='\n\n', sep='\n')



#### Make a df that has camera locations and intersections
Intersections are present (and addresses), but we do not have lat/long info for all cams

In [ ]:
# we had some incorrect data in the code below, but have a creative fix.
cam_locs = rlc_df.groupby(['camera_id', 'intersection']).max().reset_index()
cam_locs.head()

# we find there is a mismatch between lens, one of them is duplicated
len(cam_locs)  # 364 total
len(cam_locs['camera_id'].unique()) # 363

cam_locs[cam_locs['camera_id'].duplicated()]  # 1421 is dupe
print('Two of them\n', cam_locs[cam_locs['camera_id'] == '1421'])  # we see two of them
print()

# Which one is it?
print('Damen/Diversey', rlc_df[(rlc_df['camera_id']=='1421') & (rlc_df['intersection']=='DAMEN AND DIVERSEY')]['camera_id'].count())
print('Laramie/Fullerton:', rlc_df[(rlc_df['camera_id']=='1421') & (rlc_df['intersection']=='LARAMIE AND FULLERTON')]['camera_id'].count())

# Turns out that a camera has two locations. One was only used one time.  We drop it.
cam_locs = cam_locs[(cam_locs['camera_id']!='1421') | (cam_locs['intersection']!='DAMEN AND DIVERSEY')]
print("Total cams", len(cam_locs))  # 363 total (got rid of the bad one)

In [ ]:
cam_locs.isna().sum()  # missing location for 19 cameras.  Let's fix it

Looks like we also are missing 19 of the 363 cam locations.  Let's look it up!

We actually changed this code.  We have gone to the maps to get intersection locations, but will leave old code in case we ever need exact location of cameras.

In [ ]:

'''
This section goes through all of the rlc and assigns latlong
Many lights are missing it.  
For each light, there is an address though.
We use geocoding to get the latlong
'''

# let's get all of the red light cameras with their gps location.  
# This will aid in placing the accidents at rlc intersections later (if closer than threshold point to point)
# Some RLCs are missing location data,  but have addresses.  I can use geocoding I guess to look them up.


# geolocator = Nominatim(user_agent="https://github.com/sciencelee/chicago_rlc")  # please change to match repo
 
 
# def get_geocode(lat, long, address):
#     if lat > 0:  # it's a location
#         return (lat, long)
#     else: # it's a proper location tuple, and assumed to be correct latlong
#         if address in address_fix.keys(): address = address_fix[address]  # errors in the dataset chars omitted
#         # if we make it this far, we have no record of this cam_id yet, and it doesn't have a proper location
#         location = geolocator.geocode(address + ', Chicago, IL')
#         if location == None:
#             print(address+':'+address+' : could not geolocate') # print it out if we can't find (address errors)
#         else:
#             return (location.latitude, location.longitude)

        


# Got it down to one-liner for this. Found out you can't extract and assign series like you can variables
#cam_locs['location'] = cam_locs.apply(lambda x: get_geocode(x.latitude, x.longitude, x.address), axis=1)

In [ ]:
# cam_locs['location'].head()
# cam_locs['latitude'] = cam_locs['location'].apply(lambda x: x[0])
# cam_locs['longitude'] = cam_locs['location'].apply(lambda x: x[1])

# cam_locs = cam_locs.drop(columns=['violation_date', 'violations', 'month', 'weekday', 'year', 'location'])

In [ ]:
cam_locs.info()
cam_locs.isna().sum() # No longer missing location for 19 cameras.

### Create cam_locations TABLE
from cam_locs AND cam_startend from cam_startend

In [ ]:
make_table(cam_locs, 'cam_locations', c, conn)
make_table(cam_startend, 'cam_startend', c, conn)

###  Fix lat/long data
We still have missing lat/long info for our rlc_df.  Let's fix it
Before moving on.  Now that we have cam_locs, we can fix our rlc_df

In [ ]:
rlc_df.isna().sum()  


### Change cam position to intersection lat/lon
During EDA, we found out that five cameras were in completely wrong lat/long location.  
Several others were located a little too far from the intersection to work properly.  When we rebuild the db, we will use bigger number than 30 m.

I intend to use intersection locations in lieu of camera locations.  I hope this makes all of my position data consistent for gathering crash info, and eliminates chicago data portal errors.

When using cam location, it is sometimes up to 35 m up road where cam position is.  This would cause us to misidentify crashes from other intersections or miss some in the intersection of interest. 

Remedy: Use center point of intersection for all cams.  I have done this in the intersection_chars (looked it up on google maps for all 180+ intersections)

In [ ]:
int_df.columns


In [ ]:
def location_correction(int_df, intersect, latlong):
    # lookup function from intersection df to get the lat long
    # int_df is the intersection characteristic frame from 1) above
    # intersect is the intersection name used to link tables/df
    # latlong is either 'lat' or 'long'
    if latlong == 'lat':
        lat = int_df[int_df['intersection']==intersect]['lat'].values[0]
        if lat==None: print(lat, intersect)
        return lat
    else:
        long = int_df[int_df['intersection']==intersect]['long'].values[0]
        return long

cam_locs['latitude'] = cam_locs['intersection'].apply(lambda x: location_correction(int_df, x, 'lat'))
cam_locs['longitude'] = cam_locs['intersection'].apply(lambda x: location_correction(int_df, x, 'long'))

In [ ]:
rlc_df.intersection.head()

In [ ]:
int_df[int_df['intersection']=='IRVING PARK AND KILPATRICK']

Make my cameras have a location that is center of intersection instead of exact cam location.

In [ ]:
#⏳⏳⏳⏳⏳  (8min on my macbook pro)
def read_loc(int_df, intersection):
    # This function looks up the new camera/intersction location using the intersection name as key
    cam = int_df[int_df['intersection']==intersection]
    #print(cam)
    return (float(cam['lat']), float(cam['long']))
        


# create a location column so we only have to do it once
rlc_df['location'] = rlc_df['intersection'].apply(lambda x: read_loc(int_df, x))
rlc_df.head()

In [ ]:
# then add in the new lat longs to the df
rlc_df['latitude'] = rlc_df['location'].apply(lambda x: x[0])
rlc_df['longitude'] = rlc_df['location'].apply(lambda x: x[1])

In [ ]:
rlc_df[rlc_df.latitude.isna()]['intersection'].unique()  # which intersections am I still missing

In [ ]:
rlc_df.head()

In [ ]:
# get rid of location column.  We have latitude and longitude in separate columns and don't want BLOBs for now
if 'location' in rlc_df.columns:
    rlc_df.drop(columns=['location'], inplace=True)

### Create daily_violations TABLE
Now we have gone back to rlc_df and fixd the location data to line up with intersections.
We commit it to our database.

In [ ]:
make_table(rlc_df, 'daily_violations', c, conn)

## 4) Build intersection_cams TABLE
Create a table to store intersection and camera data using int_cams df.

### Preprocess intersection_cams
We will now focus on trying to bring rlc intersections to our crashes
We find that we have 363 cameras at 183 intersections.

Group my data by intersection and pull out the individual camera_id.  We place them into separate columns.  This allows us to store the individual camera ids associated with each intersection should we need them for a query.

In [ ]:
int_cams = cam_locs.groupby(['intersection']) \
                    .agg({'latitude':pd.Series.max, 'longitude':pd.Series.max,}) \
                    .reset_index()

int_cams['cam1'] = int_cams['intersection'] \
                            .apply(lambda x: cam_locs[cam_locs['intersection']==x]['camera_id'].iloc[0])

int_cams['cam2'] = int_cams['intersection'].apply( \
                            lambda x: None if len(cam_locs[cam_locs['intersection']==x])==1 \
                            else cam_locs[cam_locs['intersection']==x]['camera_id'].iloc[1])

int_cams['cam3'] = int_cams['intersection'].apply( \
                            lambda x: None if len(cam_locs[cam_locs['intersection']==x])<3 \
                            else cam_locs[cam_locs['intersection']==x]['camera_id'].iloc[2])                             

int_cams.head()

In [ ]:
print('Total Cameras', len(cam_locs))
print('Total Intersections', len(int_cams))

### Create intersection_cams TABLE 
Create it from int_cams

In [ ]:
make_table(int_cams, 'intersection_cams', c, conn)

## 5) Create signal_crashes and all_crashes TABLE
Create all_crashes from crash_df.  All crashes will include all signal crashes in the city.

Create signal_crashes from crash_df. This will only be the signal crashes at red light cam intersections.

The crash data has a column for 'intersection related crashes', and another for 'traffic signal'.  We use this to filter our data.


### Crash data preprocessing


In [ ]:
# Crash Data
crash_data = client.get("85ca-t3if", 
                     where="crash_date > \'2016-01-01T00:00:00.000\'",
                     limit=1000000,
                    )

crash_df = pd.DataFrame.from_records(crash_data) # Convert to pandas DataFrame

In [ ]:
# drop a few columns we don't need, including location (we have lat/long)
dropme = ['statements_taken_i', 'private_property_i', 'photos_taken_i', 'dooring_i', 'date_police_notified','location']

crash_df.drop(columns=dropme, inplace=True)

In [ ]:
crash_df.isna().sum()

We have 2.5k entries that have no location.  Let's drop them

In [ ]:
crash_df.dropna(subset=['latitude',], inplace=True)  # get rid of na locations

### Contents of crash data

Let's look at what is in the data.  We have over 30 columns.  Much of it is categorical.  We want to look and see what is in here before we continue.  

We start with over half a million crashes.

In [ ]:
# What's in this data?
col_interest = ['traffic_control_device', 'device_condition', 'weather_condition',
       'lighting_condition', 'first_crash_type', 'trafficway_type',
       'alignment', 'roadway_surface_cond', 'road_defect', 'report_type',
       'crash_type', 'hit_and_run_i', 'damage', 'prim_contributory_cause',
       'sec_contributory_cause', 'street_no', 'street_direction',
       'street_name', 'beat_of_occurrence', 'num_units', 'most_severe_injury', 
        'injuries_fatal', 'injuries_incapacitating',
       'injuries_non_incapacitating', 'injuries_reported_not_evident',
       'injuries_no_indication', 'injuries_unknown', 'crash_hour',
       'crash_day_of_week', 'crash_month', 'latitude', 'longitude', 'lane_cnt',
       'intersection_related_i', 'crash_date_est_i',
       'work_zone_i', 'work_zone_type',
       'workers_present_i']

for col in col_interest:
    print(col, crash_df[col].unique())

### Create all_crashes TABLE
We create this and will continue to process it to deliver a table with only crashes at signaled intersections.

In [ ]:
make_table(crash_df, 'all_crashes', c, conn)

### Filter for desired crashes (intersections with signal)
The contents of crash data section has helped us.  
We can filter 'traffic_control_device' == 'TRAFFIC SIGNAL'.  
We can filter 'intersection_related_i' == 'Y'

This will leave us with only crashes that occurred at/because of intersections, and with a signal at the intersection.

intersection_related_i: A field observation by the police officer whether an intersection played a role in the crash. Does not represent whether or not the crash occurred within the intersection.

In [ ]:
crash_df = crash_df[(crash_df['traffic_control_device']=='TRAFFIC SIGNAL') & \
                    (crash_df['intersection_related_i']=='Y')]

### Add intersection to my signal crashes

Look up each lat long for crash and get the corresponding intersection if we have it.

Tried to look up using geo equations took forever.  Also slow using pythag thrm, took long time.  Finally settled on using a simple box.  I basically check to see if the crash location is inside a box defining the intersection (within 50m of the intersection centerpoint)

In [ ]:
#⏳⏳⏳⏳⏳
# This takes too long to process.  Let's simplify it and make it a box instead.
box_side = 100  # effectively makes it check for crash being within 40m of intersection
box_lat = box_side / 111070 / 2 # 111070 is meters in deg lat in Chicago
box_long = box_side / 83000 / 2 # 83000 is meters in deg long in Chicago

def box_check(lat, long, int_df):
    answer = (int_df[  (int_df['lat'] > (lat - box_lat)) & 
                      (int_df['lat'] < (lat + box_lat)) &
                      (int_df['long'] > (long - box_long)) &
                      (int_df['long'] < (long + box_long))
                     ])
    if answer.empty: return None
    return answer['intersection'].values[0]
    
# THIS SEEMS TO WORK WITH SPEED AND ELIMINATES MEMORY PROBLEM
crash_df['intersection'] = crash_df.apply(lambda x: box_check(float(x.latitude), 
                                                              float(x.longitude), 
                                                              int_df), axis=1)

Extract the integer date data from the crash_date column

In [ ]:
# #Ex: df[['two', 'three']] = df[['two', 'three']].astype(float)
crash_df['crash_date'] = pd.to_datetime(crash_df['crash_date'])
crash_df['year'] = crash_df['crash_date'].apply(lambda x: int(x.year))
crash_df['month'] = crash_df['crash_date'].apply(lambda x: int(x.month))
crash_df['day'] = crash_df['crash_date'].apply(lambda x: int(x.day))
crash_df['hour'] = crash_df['crash_date'].apply(lambda x: int(x.hour))

In [ ]:
crash_df.info()

### Create signal_crashes TABLE
Create from crash_df.  These 60k crashes are labeled with region_id and intersection as foreign keys for SQL queries.

In [ ]:
make_table(crash_df, 'signal_crashes', c, conn)

## 6) Build hourly_congestion TABLE 

Chicago tracks periodic (irregular) bus speed data.  The data is aggregated so that we get an average bus_speed for every region at least once an hour while buses are running.  

Build from all_traffic DataFrame.

For this one, we have to combine two different datasets.  Chicago changed the way data was recorded in 2018.  Columns are similar, but more data collected.

### First dataset
Won't be able to table it until we get both datasets.

In [ ]:
# Congestion Data
traffic_df = client.get("emtn-qqdi", 
                     #where="TIME > \'2015-01-01T00:00:00.000\'",
                     where='TIME > \'2016-01-01T00:00:00.000\'',
                     limit=10000000,
                    )

traffic_df = pd.DataFrame.from_records(traffic_df) # Convert to pandas DataFrame

In [ ]:
traffic_df.rename(columns={'number_of_reads':'num_reads'}, inplace=True)
traffic_df['time'] = pd.to_datetime(traffic_df['time'])
traffic_df['bus_count'] = traffic_df['bus_count'].astype(int)
traffic_df['num_reads'] = traffic_df['num_reads'].astype(int)
traffic_df['speed'] = traffic_df['speed'].astype(float)

### Second dataset

In [ ]:
# Congestion data from later
traffic_df2 = client.get("kf7e-cur8", #2018 to present
                     select='time, region_id, speed, bus_count, num_reads',  # this set is huge, so we won't get all       
                     where="TIME < \'2021-01-01T00:00:00.000\'",
                     limit=10000000,
                    )

# Convert to pandas DataFrame
traffic_df2 = pd.DataFrame.from_records(traffic_df2)


In [ ]:
#traffic2_df.rename(columns={'number_of_reads':'num_reads'}, inplace=True)
traffic_df2['time'] = pd.to_datetime(traffic_df2['time'])
traffic_df2['bus_count'] = traffic_df2['bus_count'].astype(int)
traffic_df2['num_reads'] = traffic_df2['num_reads'].astype(int)
traffic_df2['speed'] = traffic_df2['speed'].astype(float)

### Preprocess hourly_congestion
We have two separate traffic_dfs.  There is data prior to 2018 and after in two different api endpoints.


In [ ]:
traffic_df.head()
traffic_df2.head()
traffic_df2.info()
print()
traffic_df.info()

In [ ]:
# Merge my two data sets for congestion by region
all_traffic = pd.merge(traffic_df, traffic_df2, how='outer')
print('traffic dfs merged')

In [ ]:
all_traffic['hour'] = all_traffic['time'].dt.hour
print('added hour column')

all_traffic['day'] = all_traffic.time.dt.day
print('added day column')

all_traffic['month'] = all_traffic.time.dt.month
print('added month column')

all_traffic['year'] = all_traffic.time.dt.year
print('added year column')

all_traffic['weekday'] = all_traffic.time.dt.weekday
print('added weekday column')

In [ ]:
print(len(all_traffic))  # lots of dupes 
all_traffic = all_traffic.groupby(['year', 'month', 'day', 'hour', 'region_id']).mean().reset_index()
all_traffic.info()

In [ ]:
all_traffic.head()

In [ ]:
# couple minutes
make_table(all_traffic, 'hourly_congestion', c, conn)

### Speed fix for congestion
Congestion is measured by average bus speed.

The problem:
- Overnight (between 11 and 5am) we have few bus routes.
- Some regions have no buses overnight
- Some regions have only a few buses 
- Some buses are ending routes and have only a few reads
- Some buses are stationary (next morning staging)

The fix:
- replace speed for few buses/reads if speed is low
- we assume low buses/reads to be overnight when congestion is minimal
- replacement speed is a low congestion quantile speed (90% or so)

In [ ]:
## Let's get the 0.90 quantile for every region, and then use that to fill in missing data

regions_90 = all_traffic.groupby(['region_id'])['speed'].quantile(0.9).reset_index()


In [ ]:
regions_90.head()


In [ ]:
#### 5 MINUTES OR SO

#my read on this is that few buses run 24/7, so the data is unreliable.  
# buses stage for next morning.  You can see them all along Clark, LSD etc.  
# They have speed=0 and may be recording.  Could talk to owner of dataset.

# I will draw the cutoff at 100 reads, 5 buses, speed < 10
# in that case I will put in a quantile speed for the region


def speed_check(bus, speed, reads, region_id, regions_90):
    if (bus <= 5 or reads < 100) and speed < 25 or speed > 40:
        return regions_90[regions_90['region_id']==region_id]['speed'].values[0]
    else:
        return speed
    

# apply is SLOOOOOOWWW, but not sure how else to accomplish this without iterating
all_traffic['speed'] = all_traffic.apply(lambda x: speed_check(x.bus_count, x.speed, x.num_reads, x.region_id, regions_90), axis=1)
      


### Create hourly_congestion TABLE

In [ ]:
make_table(all_traffic, 'hourly_congestion', c, conn)

#### Add quantile speed
In models, congestion was not as useful as the EDA suggested.  I would like to create a quantile speed for each area.  That way downtown traffic midday is not worse than peak traffic in other regions.





In [ ]:
import timeit

# This is slow!  Original speed took about 0.01s per iteration!  (200 minutes!!!!)
# used dictionary of smaller dfs and looked up inside lambda function (0.003 per iteration) (60 minutes!!!)

# split into 29 dfs, cycle perform lambda on each (no need for df.apply x, could do series.apply) (2 minutes...)

def quant_speed(speed, df):
    # returns a percentile speed for the region
    quant = percentileofscore(df['speed'], speed)
    return quant
    
    
new_df = pd.DataFrame(columns=all_traffic.columns)
new_df['quantile_speed'] = None

for i in all_traffic.region_id.unique():
    df = all_traffic[all_traffic['region_id'] == str(i)]
    df['quantile_speed'] = df['speed'].apply(lambda x: quant_speed(x, df))
    new_df = new_df.append(df)
    
new_df.head()
# filtering first and pull from dict increased speed by 15x by using dictionary lookup
# starttime = timeit.default_timer()
# all_traffic['quantile_speed'] = all_traffic.apply(lambda x: quant_speed(x.speed, x.region_id, df_dict[x.region_id]), axis=1)
# print("The time difference is :", timeit.default_timer() - starttime)

In [ ]:
new_df.region_id.unique()

In [ ]:
make_table(new_df, 'hourly_congestion', c, conn)

## 7) Create hourly_weather 
Created from wx_df dataframe.  

This data is hourly historical data from a bulk query from openweathermap.org for the city of Chicago.
Columns of interest:
- temp (deg K)
- rain_1h (mm rain in last hour)
- snow_1h (mm snow in last hour)
- weather_main (description)


### Preprocess hourly_weather

In [ ]:
# Import weather data from csv
wx_df = pd.read_csv('data/chi_wx.csv')

In [ ]:
wx_df.weather_main.unique()

In [ ]:
wx_df['time'] = pd.to_datetime(wx_df['dt_iso'].apply(lambda x: x[:-4]))
wx_df.head()
wx_df.info()

In [ ]:
wx_df['rain_3h'] = wx_df['rain_3h'].fillna(0)
wx_df['rain_1h'] = wx_df['rain_1h'].fillna(0)
wx_df['snow_3h'] = wx_df['snow_3h'].fillna(0)
wx_df['snow_1h'] = wx_df['snow_1h'].fillna(0)
wx_df['temp'] = wx_df['temp_max']
wx_df['year'] = wx_df.time.dt.year
wx_df['month'] = wx_df.time.dt.month
wx_df['day'] = wx_df.time.dt.day
wx_df['hour'] = wx_df.time.dt.hour
wx_df['weekday'] = wx_df.time.dt.weekday

In [ ]:
wx_df.describe()

In [ ]:
try:
    wx_df = wx_df.drop(columns=['dt', 
                        'dt_iso', 
                        'timezone', 
                        'city_name', 
                        'lat', 
                        'lon', 
                        'feels_like', 
                        'temp_min', 
                        'temp_max',
                        'pressure',
                        'sea_level',
                        'grnd_level',
                        'humidity',
                        'wind_speed',
                        'wind_deg',
                        'clouds_all',
                        'weather_description',
                        'weather_icon',
                        'weather_id',
                       ], axis=1)
except:
    print('Failed')

In [ ]:
print(len(wx_df))
print(wx_df.duplicated().sum())


print('Total hours in 6 years:', 365.25 * 24 * 6)
print('Unique entries:', len(wx_df.drop_duplicates()))  
# missing a few entries (700+ out of 52k)  Am I missing a month??

print()
print(wx_df.time.min(), wx_df.time.max())  # OH!!!!  I am missin last month
print('Total hours in 6 years (-1 mos):', 365.25 * 24 * 6 - 31 * 24)  # okay, we are only missing a few


wx_df.drop_duplicates(inplace=True)

### Create hourly_weather TABLE

In [ ]:
make_table(wx_df, 'hourly_weather', c, conn)

## 8) Create region_data TABLE 
Created from region_df dataframe.

Wanted to add a table which contained the region information.  Congestion TABLE uses region_id to break city into 29 traffic regions.  We use the regions for congestion (bus speed) data.

### Preprocess region_data TABLE

In [ ]:
# THis time we only grab what we need

region_df = client.get("kf7e-cur8", # regional congestion current data
                         select='region_id, region, description, north, south, east, west',
                         limit=1000
                    )

# Convert to pandas DataFrame
region_df = pd.DataFrame.from_records(region_df)  # should only return most recent for each region

In [ ]:
region_df = region_df.groupby('region_id').max().reset_index()

In [ ]:
# need these as floats so we can compare them
region_df[['north', 'south', 'east', 'west']] = region_df[['north', 'south', 'east', 'west']].astype(float)
region_df.info()


### Add region to my crash df

In [ ]:
crash_df[['latitude', 'longitude']] = crash_df[['latitude', 'longitude']].astype(float)
crash_df.info()

In [ ]:
# add in the region for my crashes
# Resource hog
crash_df.columns


def which_region(lat, long, region_df):
    #print(lat, long)
    row = region_df[(region_df['east'] >= long) &
                    (region_df['west'] < long) &
                    (region_df['north'] >= lat) &
                    (region_df['south'] < lat)]['region_id'].max()
    return row

#df.iloc[:5]
# takes some 5min
crash_df['region_id'] = crash_df.apply(lambda x: which_region(x.latitude, x.longitude, region_df), axis=1)

In [ ]:
len(crash_df)
crash_df.columns

crash_df['time'] = pd.to_datetime(crash_df.crash_date)
crash_df['year'] = crash_df.time.dt.year
crash_df['month'] = crash_df.time.dt.month
crash_df['day'] = crash_df.time.dt.day
crash_df['hour'] = crash_df.time.dt.hour
crash_df['weekday'] = crash_df.time.dt.weekday

### Create region_df TABLE and update signal_crashes TABLE

With new info added to crash table, we overwrite previous table.

In [ ]:
make_table(region_df, 'region_data', c, conn)
print()
make_table(crash_df, 'signal_crashes', c, conn)  # also update my crash data

### Add region_id to intersection_cams
While I'm here and I have the function ready.
I would like to add region_id number to my red light camera (daily_violations TABLE)
The region there will help me link the daily_violations and hourly_congestion TABLEs

*** NOTE: Makes more sense to come back and put the region into the intersection_cameras table to speed this up

In [ ]:
# 1 minutes
#rlc['region_id'] = 
int_cams['region_id'] = int_cams.apply(lambda x: which_region(x.latitude, x.longitude, region_df), axis=1)


In [ ]:
## commit my change
make_table(int_cams, 'intersection_cams', c, conn)

<b> Use this code to test any of your tables for proper data storage </b>

In [ ]:
query = c.execute("SELECT camera_id, violations FROM daily_violations;").fetchall()
print(query[:5])
print(len(query))

### Add intersections to signal_crashes 
Want to add this as a foreign key to help with queries.

In [ ]:
sql_fetch_tables(c, conn)

In [ ]:
# read data back in to prevent having to rerun code.
df = pd.read_sql_query("SELECT * FROM signal_crashes", conn)
camloc_df = pd.read_sql_query('SELECT * FROM cam_locations', conn)
ints_df = pd.read_sql_query('SELECT * FROM intersection_cams', conn)


In [ ]:
#ints_df.astype({'longitude':float})
pd.options.display.max_rows = 200


### Add intersection data to crashes

In [ ]:
# Let's simplify it and make it a box instead.
box_side = 100  # effectively makes it check for crash being within 25m of interscection
box_lat = box_side / 111070 / 2 # 111070 is meters in deg lat in Chicago
box_long = box_side / 83000 / 2 # 83000 is meters in deg long in Chicago

def box_check(lat, long, int_df):
    n = lat + box_lat
    s = lat - box_lat
    e = long + box_long
    w = long - box_long
    # print('n', n, 's', s, 'e', e, 'w', w, 'lat:', lat, 'long:', long)
    answer = int_df[  (int_df['lat'] > s) &
                      (int_df['lat'] < n) &
                      (int_df['long'] > w) &
                      (int_df['long'] < e)
                      
                     ]
    if answer.empty: return None
    return answer['intersection'].values[0]

    
# THIS SEEMS TO WORK AT SPEED AND ELIMINATES MEMORY PROBLEM
# this code is just to test out a chunk of data
for i in range(5000, 5100): 
    lat = float(df.iloc[i]['latitude'])
    long = float(df.iloc[i]['longitude'])
    n = lat + box_lat
    s = lat - box_lat
    e = long + box_long
    w = long - box_long
    answer = int_df[  (int_df['lat'] > s) &
                      (int_df['lat'] < n) &
                      (int_df['long'] > w) &
                      (int_df['long'] < e)]['intersection'].values
    if len(answer): print(answer[0])
    
# 99th Halsted: 41.714230	-87.643043
# MOMENT OF TRUTH
df['intersection'] = df.apply(lambda x: box_check(float(x.latitude), float(x.longitude), int_df), axis=1)



### Overwrite signal_crashes TABLE
Now it is updated with intersections

In [ ]:
make_table(df, 'signal_crashes', c, conn)


## 9) Create all_hours TABLE
Make a table that just has every date and hour for every intersection.  Will help out my queries so I can use LEFT JOINS only.  Will use the weather data (hourly) to build my table.

In [ ]:
#int_chars = pd.read_sql_query("SELECT * FROM intersection_chars", conn)
v_df = pd.read_sql_query("SELECT * FROM daily_violations", conn)
wx_df = pd.read_sql_query("SELECT * FROM hourly_weather", conn)
rlc_df = pd.read_sql_query("SELECT * FROM daily_violations", conn)

In [ ]:
grouped_dates = wx_df.groupby(['year', 'month', 'day', 'hour']).max().reset_index()[['year', 'month', 'day', 'hour']]
grouped_dates = grouped_dates[grouped_dates['year']>2015]

In [ ]:
# didn't work 
grouped_dates.head()
grouped_dates.info()

In [ ]:
big_df = grouped_dates.copy()

big_df['datetime'] = big_df.apply(lambda x: datetime(x.year, x.month, x.day, x.hour), axis=1)

In [ ]:
my_ints = rlc_df.intersection.unique()

In [ ]:
# Should only have to do this once
#for i in range(len(wx_df)):
big_df['intersection'] = my_ints[0]


for i in range(1, len(my_ints)):
    df = grouped_dates.copy()
    df['intersection'] = my_ints[i]
    big_df = pd.concat([big_df, df])



In [ ]:
big_df.intersection.isna().sum()

In [ ]:
big_df.intersection.unique()
big_df.dropna(subset=['intersection'], inplace=True)


In [ ]:
big_df.info()

In [ ]:
big_df.head()

### Create all_hours TABLE

In [ ]:
make_table(big_df, 'all_hours', c, conn)

## 10) Build int_startend TABLE 

Create a table which serves to provide start and end dates for camera functioning at every intersection.

We will just look at the earliest and latest entries from the violations.
When cameras are off, they just disappear from the database.

All cameras have gaps (maintenance and malfunctions). I do not differentiate these in this project.

### Preprocess int_startend
Add intersection start-end dates

In [ ]:
v_df = pd.read_sql_query("SELECT * FROM daily_violations", conn)

In [ ]:
v_df['start'] = v_df['violation_date']
v_df['end'] = v_df['violation_date']

start_end = v_df.groupby('intersection').agg({'start':'min', 'end':'max'}).reset_index()

In [ ]:
start_end.start.sort_values().tail(50)
start_end.head()

### Create start_end

In [ ]:
make_table(start_end, 'int_startend', c, conn)

## Add speed to int_char TABLE
Use the average (or mode speed) from crashes at intersection to add to my intersection charateristics.

In [ ]:
speed_df = pd.read_sql_query("SELECT * FROM signal_crashes", conn)
int_char_df = pd.read_sql_query("SELECT * FROM intersection_chars", conn)


In [ ]:
speed_df.posted_speed_limit = speed_df.posted_speed_limit.astype(int)

In [ ]:
speed_df = speed_df.groupby('intersection').agg({'posted_speed_limit':max}).reset_index()  # mode is from scipy

In [ ]:
speed_df.posted_speed_limit.unique()
speed_df.head()

In [ ]:

# df is defined globally
def speed_lookup(intersection):
    # needed to put a try in there because one intersection had no crashes over time period
    try:
        speed = speed_df[speed_df['intersection']==intersection]['posted_speed_limit'].values[0]
    except:
        speed=30
    return speed
    
int_char_df['speed'] = int_char_df['intersection'].apply(speed_lookup)

In [ ]:
int_char_df.head()

### Rewrite int_char

In [ ]:
make_table(int_char_df, 'intersection_chars', c, conn)

### Add Red Light On/Off to crash data

In [ ]:
crash_df2 = pd.read_sql_query("SELECT * FROM signal_crashes", conn)
#se_df = pd.read_sql_query("SELECT * FROM int_startend", conn)

In [ ]:
#crash_df2 = crash_df.merge(se_df, on='intersection', how='left')


In [ ]:
crash_df2.columns

In [ ]:
crash_df2['start'] = pd.to_datetime(crash_df2['start'])
crash_df2['end'] = pd.to_datetime(crash_df2['end'])
crash_df2['crash_date'] = pd.to_datetime(crash_df2['crash_date'])

In [ ]:
# need to determine if crash occurred in or outside of cam on dates

def rlc_state(start, end, my_date):

    if (end - my_date).days >= 0 and (my_date - start).days >= 0:
        return 1
    elif (my_date - end).days > 0:
        return 0
    elif (start - my_date).days > 0:
        return 0
    else:
        return None

crash_df2['rlc_state'] = crash_df2.apply(lambda x: rlc_state(x.start, x.end, x.crash_date), axis=1)

In [ ]:
print('Total crashes rlc on: {}'.format(crash_df2['rlc_state'].sum()))
print('Total crashes rlc off: {}'.format(crash_df2[crash_df2['rlc_state']==0]['crash_date'].count()))


In [ ]:
make_table(crash_df2, 'signal_crashes', c, conn)

## Add camera state to all_hours
Needed to do this to make my queries easier since I don't have an outer 
join in SQLite3.  This takes a significant amount of time.  Hours 

In [ ]:

all_hours = pd.read_sql_query("SELECT * FROM all_hours", conn)


In [ ]:
#all_hours['datetime'] = pd.to_datetime(all_hours)
all_hours.head()

In [ ]:
all_hours.datetime = all_hours.apply(lambda x: datetime(int(x.year), x.month, x.day, x.hour), axis=1)

In [ ]:
all_hours.tail()

In [ ]:
se_df = pd.read_sql_query("SELECT * FROM int_startend", conn)
se_df['start'] = pd.to_datetime(se_df['start'])
se_df['end'] = pd.to_datetime(se_df['end'])

se_df.intersection

se_dict = {x:[se_df[se_df['intersection']==x]['start'].values[0],  
              se_df[se_df['intersection']==x]['end'].values[0]] for x in se_df.intersection}


In [ ]:
# Let's make it faster
def get_camstate2(date, start, end):
    if date >= start and date <= end:
        return 1
    else:
        return 0

    
intersections = list(se_dict.keys())
intersection = intersections[0]
new_df = all_hours[all_hours['intersection']==intersection]
new_df['rlc_state'] = new_df['datetime'].apply(lambda x: get_camstate2(x, 
                                                  se_dict[intersection][0],
                                                  se_dict[intersection][1]
                                                    ))
print(new_df.head())
print(intersection, 'complete')

for intersection in intersections[2:]:
    concat_me = all_hours[all_hours['intersection']==intersection]
    concat_me['rlc_state'] = concat_me['datetime'].apply(lambda x: get_camstate2(x, 
                                                  se_dict[intersection][0],
                                                  se_dict[intersection][1]
                                                    ))
    print(intersection, end=',')
    new_df = pd.concat([new_df, concat_me])

print("END")

The above code was extremely slow until I broke up the df into parts in a loop. 

In [ ]:
new_df.head()

In [ ]:
new_df.isna().sum()

In [ ]:
new_df.rlc_state.sum()

In [ ]:
make_table(new_df, 'rlc_all_hours', c, conn)


## Adding COVID rates

In [ ]:
daily_covid= client.get("naz8-j4nc", 
                     limit=20000,
                    )

daily_covid = pd.DataFrame.from_records(daily_covid) # Convert to pandas DataFrame

In [ ]:
make_table(daily_covid, 'daily_covid', c, conn)

## Close my connection to the db

In [ ]:
c.close()

In [ ]:
new_df.rlc_state.sum()